### PR TITLE
[Feat] Add support for enabling inference

### PR DESCRIPTION
**Pull request by tchaton**
### Description

Fixes #[ISSUE NUMBER]

Replace this line by a one sentence summary of your PR.

If necessary, use the following space to provide context or more details.

### Contribution Checklist

If your contribution modifies code in the core library (not docs, tests, or examples), please fill the following checklist.

- [ ] My contribution is listed in CHANGELOG.md with attribution.
- [x] My contribution modifies code in the main library.
- [ ] My modifications are tested.
- [ ] My modifications are documented.

#### Optional

If you make major changes to the core library, please run `make alltests` and copy-paste the content of `alltests.txt` below.

~~~shell
[PASTE HERE]
~~~

### CODE REVIEW

File: learn2learn/utils/lightning.py [hunk 1]
```diff
@@ -3,17 +3,140 @@
 """
 Some utilities to interface with PyTorch Lightning.
 """
-
+from typing import Optional, Callable
+import learn2learn as l2l
 import pytorch_lightning as pl
+from torch.utils.data._utils.worker import get_worker_info
+from torch.utils.data import IterableDataset, Dataset
+from torch.utils.data._utils.collate import default_collate
```
Comment (pietrolesci) [inline learn2learn/utils/lightning.py:11]: It `default_collate` needed?

File: learn2learn/utils/lightning.py [hunk 2]
```diff
@@ -3,17 +3,140 @@
 """
 Some utilities to interface with PyTorch Lightning.
 """
-
+from typing import Optional, Callable
+import learn2learn as l2l
 import pytorch_lightning as pl
+from torch.utils.data._utils.worker import get_worker_info
+from torch.utils.data import IterableDataset, Dataset
```
Comment (pietrolesci) [inline learn2learn/utils/lightning.py:10]: Is `Dataset` needed?

Comment (tchaton) [reply learn2learn/utils/lightning.py]: Done !

File: learn2learn/utils/lightning.py (hunk 1 above)
Comment (tchaton) [reply learn2learn/utils/lightning.py]: Done !

File: learn2learn/algorithms/lightning/lightning_episodic_module.py [hunk 3]
```diff
@@ -143,3 +161,4 @@ def configure_optimizers(self):
             gamma=self.scheduler_decay,
         )
         return [optimizer], [lr_scheduler]
+
```
Comment (seba-1511) [inline learn2learn/algorithms/lightning/lightning_episodic_module.py:164]: Remove new line. (Check locally with `make lint`)

File: learn2learn/utils/lightning.py [hunk 4]
```diff
@@ -32,38 +154,22 @@ def __init__(
...
-                return self.tasks.sample()
-
-            def __len__(self):
-                return self.length
-
-        return Epochifier(taskset, epoch_length)
-
     def train_dataloader(self):
-        return EpisodicBatcher.epochify(
+        return Epochifier(
```
Comment (seba-1511) [inline learn2learn/utils/lightning.py:158]: Was `Epochifier` inadvertently removed? Or should it be replaced by `TaskDataParallel`?

Also, if I remember correctly `Epochifier` was preventing us from using pytorch_lightning > 1.0.2. I believe the fix was just to inherit from a data class in lightning. @nightlessbaron do you remember which class and if that was enough?

File: learn2learn/utils/lightning.py [hunk 5]
```diff
@@ -3,17 +3,139 @@
...
+        collate_fn: Optional[Callable] = None
+    ):
+        """
+        This class is used to sample epoch_length tasks to represent an epoch.
+
+        It should be used when using DataParallel
+
+        Args:
+            taskset: Dataset used to sample task.
+            epoch_length: The expected epoch length. This requires to be divisible by (num_workers * world_size).
```
Comment (seba-1511) [inline learn2learn/utils/lightning.py:30]: Second sentence of the description is not for `TaskDataParallel`.

File: learn2learn/algorithms/lightning/lightning_protonet.py [hunk 6]
```diff
@@ -139,3 +146,19 @@ def meta_learn(self, batch, batch_idx, ways, shots, queries):
         eval_loss = self.loss(logits, query_labels)
         eval_accuracy = accuracy(logits, query_labels)
         return eval_loss, eval_accuracy
+
+    def cache_on_validate_step(self, batch, batch_idx):
+        data, labels = batch
+        embeddings = self.features(data)
+        for e, l in zip(embeddings, labels):
+            self.support.append(e)
+            self.support_labels.append(l)
```
Comment (seba-1511) [inline learn2learn/algorithms/lightning/lightning_protonet.py:155]: This is potentially confusing: we're accumulating support samples across different tasks so that we can fit the classifier on increasingly more validation data. Fine, so long as each task's samples are labelled consistently with previous tasks' labels (eg, dogs are all labelled as class # 4). But that's not always the case in few-shot classification, where dogs could be class # 4 in the first task, but class # 1 in the second. Is there a way to handle both scenarios?

File: learn2learn/algorithms/lightning/lightning_episodic_module.py [hunk 7]
```diff
@@ -92,26 +102,34 @@ def training_step(self, batch, batch_idx):
...
+            )
+            self.log(
+                "valid_accuracy",
+                valid_accuracy.item(),
+                on_step=False,
+                on_epoch=True,
+                prog_bar=True,
+                logger=True,
+            )
+            return valid_loss.item()
```
Comment (seba-1511) [inline learn2learn/algorithms/lightning/lightning_episodic_module.py:128]: Do we still reach this branch when validating during training? (I'm not sure when `TrainerFn.VALIDATING` is `True` or `False`.)

**Review by seba-1511 — COMMENTED** [review]
Thanks a lot for the PR @tchaton, and for the preliminary review @pietrolesci.

Some questions and minor remarks:

* Do those changes remove the strict requirement for pytorch_lightning==1.0.2? (If yes, to be changed in lightning_episodic_module.py:16).
* List contributions (crediting to yourself) in CHANGELOG.md?

File: learn2learn/utils/lightning.py (hunk 4 above)
Comment (pietrolesci) [reply learn2learn/utils/lightning.py:158]: @seba-1511, quick answer to your first question: `Epochifier` is replaced by `TaskDataParallel`

File: learn2learn/algorithms/lightning/lightning_episodic_module.py [hunk 8]
```diff
@@ -92,26 +102,34 @@ def training_step(self, batch, batch_idx):
...
-            valid_accuracy.item(),
-            on_step=False,
-            on_epoch=True,
-            prog_bar=True,
-            logger=True,
-        )
-        return valid_loss.item()
+        if self.should_fit_on_validate:
+            # used for the algorithm to store the supports data
+            self.cache_on_validate_step(batch, batch_idx)
```
Comment (pietrolesci) [inline learn2learn/algorithms/lightning/lightning_episodic_module.py:107]: @seba-1511 `cache_on_validate_step` is implemented in the children class. For clarity, why not adding it here to the parent as "not implemented"?

File: learn2learn/algorithms/lightning/lightning_protonet.py (hunk 6 above)
Comment (pietrolesci) [reply learn2learn/algorithms/lightning/lightning_protonet.py:155]: To add to @seba-1511's comment, I think we do not need to cache data and it can potentially be incorrect from an implementation point of view. Though, I might be missing something

File: learn2learn/utils/lightning.py (hunk 4 above)
Comment (nightlessbaron) [reply learn2learn/utils/lightning.py:158]: @seba-1511, as @pietrolesci mentioned, in the benchmark, we replaced `Epochifier` with `TaskDistributedParallel`. The downside is windows users can't make use of it as stated in [#5358](https://github.com/PyTorchLightning/pytorch-lightning/issues/5358). 

> I believe the fix was just to inherit from a data class in lightning.
I couldn't figure it out. That is still unresolved.

File: learn2learn/algorithms/lightning/lightning_episodic_module.py (hunk 7 above)
Comment (tchaton) [reply learn2learn/algorithms/lightning/lightning_episodic_module.py:128]: This part of the code is run only when the users use `trainer.validate`. This is a hack and need to find a better solution.
Another option would be to provide 2 predict dataloaders, but there would be some hacks.

File: learn2learn/algorithms/lightning/lightning_protonet.py (hunk 6 above)
Comment (tchaton) [reply learn2learn/algorithms/lightning/lightning_protonet.py:155]: Oh interesting, I didn't think about this @seba-1511. 
@pietrolesci We are caching the values, as when doing production inference, we would get some labelled data and want to use them all to compute the protypes.
I believe we should prevent the transforms to make random class assignments when doing predictions.

File: learn2learn/algorithms/lightning/lightning_episodic_module.py (hunk 3 above)
Comment (tchaton) [reply learn2learn/algorithms/lightning/lightning_episodic_module.py:164]: Done !

File: learn2learn/utils/lightning.py (hunk 5 above)
Comment (tchaton) [reply learn2learn/utils/lightning.py:30]: Done !

File: learn2learn/algorithms/lightning/lightning_protonet.py (hunk 6 above)
Comment (tchaton) [reply learn2learn/algorithms/lightning/lightning_protonet.py:155]: Let's drop inference for now then, as it doesn't seem mature yet.

File: learn2learn/algorithms/lightning/lightning_episodic_module.py (hunk 7 above)
Comment (tchaton) [reply learn2learn/algorithms/lightning/lightning_episodic_module.py:128]: Removed ! Would be good to brainstorm together on the best way to support this.

File: learn2learn/utils/lightning.py (hunk 4 above)
Comment (tchaton) [reply learn2learn/utils/lightning.py:173]: Hey @seba-1511, yes it was inadvertently removed.

However, I added a TODO. It should be provided replaced by the TaskDataParallel, TaskDistributedDataParallel when tested on your side.

I will copy those files on the Flash side to no block the integration until the next learn2learn release.

Best,
T.C
